### PR TITLE
Return BlockValidationState from ProcessNewBlock if CheckBlock/AcceptBlock fails

### PR DESCRIPTION
**Pull request by jnewbery**
Net processing now passes a `BlockValidationState` object into `ProcessNewBlock()`. If `CheckBlock()` or `AcceptBlock()` fails, then PNB returns to net processing without calling the (asynchronous) `BlockChecked` Validation Interface method. net processing can use the invalid `BlockValidationState` returned to punish peers.

`CheckBlock()` and `AcceptBlock()` represent the DoS checks on a block (ie PoW and malleability). Net processing wants to know about those failed checks immediately and shouldn't have to wait on a callback. Other validation interface clients don't care about net processing submitting bogus malleated blocks to validation, so they don't need to be notified of `BlockChecked`.

Furthermore, if PNB returns a valid `BlockValidationState`, we never need to try to process (non-malleated) copies of the block from other peers. That makes it much easier to move the best chain activation logic to a background thread in future work.

### CODE REVIEW

**Comment by jnewbery** [comment]
This is the first two commits from #16279. Thanks for your work, @TheBlueMatt!

**Comment by jnewbery** [comment]
I've added a commit that deduplicates post-block-checking code, which makes the main commit in this PR smaller and (I hope) easier to understand.

**Comment by jnewbery** [comment]
I've taken the final (behaviour change) commit from #16279 and PR'ed it as #17485. That gives some justification for this PR. If we get the anti-dos / mutation checks back from validation immediately, we can use that state to mark the block as no longer in-flight, rather than by indirectly checking that the block has been written to disk.

Separating into two PRs makes this easier to review. This PR should result in no behaviour changes (modulo the `BlockChecked()` callback in net processing now being called immediately instead of on the background scheduler thread), and the follow-up PR has the (minor) behaviour change along with lots of commenting and justification.

**Comment by dongcarl** [comment]
I found the last commit in this PR (71867666388bf1137b431f944fd430f262480eba) a bit hard to parse, so I split it up in my branch here: https://github.com/bitcoin/bitcoin/compare/master...dongcarl:2019-11-processnewblock-early-return-redo-last. The main change in the last commit on this branch resides in f46102be8fdd1ef8a21b821c121ed9fa027c9d32 on mine, and f46102be8fdd1ef8a21b821c121ed9fa027c9d32 also retains the same commit message. Feel free to use/use only to review/not use.

File: src/validation.h [hunk 1]
```diff
@@ -199,22 +199,33 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
...
  * specific block passed to it has been checked for validity!
  *
- * If you want to *possibly* get feedback on whether pblock is valid, you must
- * install a CValidationInterface (see validationinterface.h) - this will have
- * its BlockChecked method called whenever *any* block completes validation.
+ * Performs initial sanity checks using the provided BlockValidationState before
+ * connecting any block(s). If you want to *possibly* get feedback on whether
+ * pblock is valid beyond just cursory mutation/DoS checks, you must install
+ * a CValidationInterface (see validationinterface.h) - this will have its
+ * BlockChecked method called whenever *any* block completes validation (note
```
Comment (dongcarl) [inline src/validation.h]: Perhaps I'm understanding wrong but this shouldn't apply anymore?

**Comment by jnewbery** [comment]
Thanks for the branch @dongcarl ! I've reset to that and made a couple of minor comment changes.

File: src/validation.h (hunk 1 above)
Comment (jnewbery) [reply src/validation.h]: BlockChecked is called for any block that we try to connect to the chain. See the call in `ConnectTip()`:

- if the call to `ConnectBlock()` failed, then we'll call `BlockChecked()` with a state that is NOT `IsValid()`.
- if the call to `ConnectBlock()` succeeded, then the block has been connected and is part of the main chain, and we'll call `BlockChecked()` with a state that IS `IsValid()`.

Note that we may never even call `ConnectTip()` if we don't try to connect the block to the most work chain.

File: src/net_processing.cpp [hunk 2]
```diff
@@ -1888,16 +1888,23 @@ void static ProcessOrphanTx(CConnman* connman, std::set<uint256>& orphan_work_se
...
-        pfrom->nLastBlockTime = GetTime();
-    } else {
+    if (!state.IsValid()) {
+        // The block failed anti-dos / mutation checks. Call BlockChecked() callback here.
+        // This clears the block from mapBlockSource.
+        BlockChecked(*pblock, state, connman);
+    } else if (!new_block) {
+        // Block was valid but we've seen it before. Clear it from mapBlockSource.
         LOCK(cs_main);
         ::mapBlockSource.erase(pblock->GetHash());
```
Comment (ariard) [inline src/net_processing.cpp]: Can't this branch combined with `BlockChecked` by passing new_block as param? Even wonder if we still need `fNewBlock` in PNB, as  `BlockChecked`  erase block from `mapBlockSource` independently of validity/novelty

**Review by ariard — COMMENTED** [review]
Concept ACK, I think commit a49a153 can be made a bit better.

File: src/net_processing.cpp [hunk 3]
```diff
@@ -1888,16 +1888,23 @@ void static ProcessOrphanTx(CConnman* connman, std::set<uint256>& orphan_work_se
...
+ *  call BlockChecked() to maybe punish peer. If this is the first time we've seen the block,
  *  update the node's nLastBlockTime. Otherwise erase it from mapBlockSource.
  */
-void static BlockProcessed(CNode* pfrom, std::shared_ptr<CBlock> pblock, bool new_block)
+void static BlockProcessed(CNode* pfrom, CConnman* connman, std::shared_ptr<CBlock> pblock, BlockValidationState& state, bool new_block)
 {
-    if (new_block) {
-        pfrom->nLastBlockTime = GetTime();
-    } else {
+    if (!state.IsValid()) {
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/net_processing.cpp]: No. `BlockProcessed` is only called when `ProcessNewBlock()` returns early because the block failed `CheckBlock()` or `AcceptBlock()`. The logic you're talking about for promoting a peer to compact block announcement in `BlockChecked()` is only invoked when `BlockChecked()` is called after successfully connecting a valid block.

File: src/net_processing.cpp (hunk 2 above)
Comment (jnewbery) [reply src/net_processing.cpp]: `BlockChecked()` will only be called from `BlockProcessed()` if the block failed `CheckBlock()`/`AcceptBlock()`. This else branch is catching the case where the block didn't fail, but we've seen it before.

If you think this can be cleaned up further, perhaps we can do it in a follow-up PR.

File: src/rpc/mining.cpp [hunk 4]
```diff
@@ -784,6 +784,9 @@ static UniValue submitblock(const JSONRPCRequest& request)
     if (!new_block && accepted) {
         return "duplicate";
     }
+    if (!dos_state.IsValid()) {
+        return BIP22ValidationResult(dos_state);
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/rpc/mining.cpp]: Yes, I was planning to do this as a follow-up PR.

I've had a quick look at your branch. I think it misses breaks the 'promote to announcing compact blocks' behaviour. We'd want to move that to a different validation interface event (probably `UpdatedBlockTip()`)

File: src/net_processing.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ariard) [reply src/net_processing.cpp]: Right, right, the call to `BlockChecked` is still in `ConnectTip` path on your branch. I still think that's counter-intuitive to have both peer punishment and compact blocks promotion in one function, so would be down to split in a follow-up PR and like you said previously rely on `UpdatedBlockTip` for the latter.

File: src/rpc/mining.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ariard) [reply src/rpc/mining.cpp]: Ah yes, agree with you, better to do it in a follow-up PR to avoid breaking things inadvertently.

File: src/net_processing.cpp [hunk 5]
```diff
@@ -1888,16 +1888,23 @@ void static ProcessOrphanTx(CConnman* connman, std::set<uint256>& orphan_work_se
 }
 
 /**
- *  A block has been processed. If this is the first time we've seen the block,
+ *  A block has been processed. If the block was failed anti-dos / mutation checks, then
+ *  call BlockChecked() to maybe punish peer. If this is the first time we've seen the block,
  *  update the node's nLastBlockTime. Otherwise erase it from mapBlockSource.
```
Comment (jkczyz) [inline src/net_processing.cpp]: The documentation strikes me as repetitive with the code. Perhaps it could be phrased differently such that if the code changes the comment doesn't need to be updated.

nit: s/was failed/failed

**Review by jkczyz — COMMENTED** [review]
Concept ACK. My previous comments had been addressed in #16279. Additional commits look good. Built and ran unit and functional tests using commit 0bf87d2.

File: src/rpc/mining.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/rpc/mining.cpp]: Work in progress: https://github.com/jnewbery/bitcoin/tree/2019-11-remove-block-checked

**Comment by jnewbery** [comment]
Thanks for the review @jkczyz . I've updated the function comment to be less redundant.

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
Thanks :octopus: . I've rebased on master.

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
Oops. Bad rebase. Should be fixed now.

File: src/validation.h [hunk 6]
```diff
@@ -157,12 +157,13 @@ extern uint64_t nPruneTarget;
  * May not be called in a
  * validationinterface callback.
  *
- * @param[in]   pblock  The block we want to process.
- * @param[in]   fForceProcessing Process this block even if unrequested; used for non-network block sources and whitelisted peers.
- * @param[out]  fNewBlock A boolean which is set to indicate if the block was first received via this call
- * @returns     If the block was processed, independently of block validity
+ * @param[in]   pblock            The block we want to process.
+ * @param[out]  state             Currently unused.
```
Comment (ariard) [inline src/validation.h]: db0eecc

nit: Is there any guidelines on parameters sorting ? You can sort by param[in] then param[out]. It makes easier to reason on function. You can also precise in commit that new `dos_state` isn't used.

File: src/net_processing.cpp [hunk 7]
```diff
@@ -1332,8 +1332,11 @@ void PeerLogicValidation::UpdatedBlockTip(const CBlockIndex *pindexNew, const CB
 /**
  * Handle invalid block rejection and consequent peer banning, maintain which
  * peers announce compact blocks.
+ * Called both in case of cursory DoS checks failing (implying the peer is likely
+ * sending us bogus data) and after full validation of the block fails (which may
```
Comment (ariard) [inline src/net_processing.cpp:1336]: a85e8b7

In `ConnectTip`, we call `ConnectBlock` then `BlockChecked` callback independently of success or failure of block validation ?

File: src/validation.h [hunk 8]
```diff
@@ -147,18 +147,29 @@ extern uint64_t nPruneTarget;
...
- * validationinterface callback.
+ * If fForceProcessing is set (or fNewBlock returns true), and state.IsValid(),
+ * barring pruning and a desire to re-download a pruned block, there should
+ * never be any reason to re-ProcessNewBlock any block with the same hash.
+ *
+ * May not be called in a validationinterface callback.
  *
  * @param[in]   pblock            The block we want to process.
- * @param[out]  state             Currently unused.
+ * @param[out]  state             Only used for failures in CheckBlock/AcceptBlock. For failure in block connection,
```
Comment (ariard) [inline src/validation.h:171]: 8fc192c

Is commit message right ? It mentions "Net processing now passes a BlockValidationState object into
ProcessNewBlock()" but in fact it's passing this object to `BlockProcessed` therefore making it used.

Comment (ariard) [reply src/validation.h:171]: 8fc192c

Still on commit message, "Net processing wants to know about those failed check immediately..." Can you explain why ? Overall, PR needs better motivation, I think you can point to changes in #18963, and the [asynchronicity gain](https://github.com/bitcoin/bitcoin/pull/18963/commits/7532cf6c56d015e8bae646138d71df825d2684ab), instead of compact blocks ones.

File: src/net_processing.cpp [hunk 9]
```diff
@@ -1981,15 +1981,21 @@ void static ProcessOrphanTx(CConnman* connman, CTxMemPool& mempool, std::set<uin
...
+void static BlockProcessed(CNode* pfrom, CConnman* connman, std::shared_ptr<CBlock> pblock, BlockValidationState& state, bool new_block)
 {
-    if (new_block) {
-        pfrom->nLastBlockTime = GetTime();
-    } else {
+    if (!state.IsValid()) {
+        // The block failed anti-dos / mutation checks. Call BlockChecked() callback here.
+        // This clears the block from mapBlockSource.
+        BlockChecked(*pblock, state, connman);
+    } else if (!new_block) {
```
Comment (ariard) [inline src/net_processing.cpp:1992]: 8fc192c

How does this new check behaves with regards to write-failure, first-seen block ? In `AcceptBlock`, if `blockPos` returned by `SaveBlockToDisk` is null, an error is set and not an invalid state, so it won't be erased by above `BlockChecked` call. I think it's catch up in current code but not after changes.

File: src/validation.h [hunk 10]
```diff
@@ -147,18 +147,29 @@ extern uint64_t nPruneTarget;
...
- * install a CValidationInterface (see validationinterface.h) - this will have
- * its BlockChecked method called whenever *any* block completes validation.
+ * Performs initial sanity checks using the provided BlockValidationState before
+ * connecting any block(s). If you want to *possibly* get feedback on whether
+ * pblock is valid beyond just cursory mutation/DoS checks, you must install
+ * a CValidationInterface (see validationinterface.h) - this will have its
+ * BlockChecked method called whenever *any* block completes validation (note
+ * that any invalidity returned via state will *not* also be provided via
+ * BlockChecked). There is, of course, no guarantee that any given block which
+ * is not a part of the eventual best chain will ever be checked.
```
Comment (ariard) [inline src/validation.h:157]: 8fc192c

I think this whole comment can be made clearer. 

"You have 2 types of feedbacks:
- unconditional : submitted block validity with regards to mutation, anti-Dos is returned through `BlockValidationState`. It may be used for housekeeping as in `BlockProcessed`
- conditional : if caller subscribe to `CValidationInterface` and if submitted block passed those first checks, `ActivateBestChain`,  will be called to try to make progress towards the best known block and make it the tip of our block chain. Submitted block may be among such tip candidates. Consensus-validity feedback on best known block may be used for peer punishment and compact block promotion as in `BlockChecked`."

**Review by ariard — COMMENTED** [review]
Re-Concept ACK 50048f4

I think well-commented this PR can be made easier for review. Note, switching from checking `CheckBlock`, `AcceptBlock` returned values to `BlockValidationState` for post-processing must be heeded, see comment on write-failure, first-seen block.

**Comment by jonatack** [comment]
Concept ACK

**Comment by jonatack** [comment]
Concept ACK, will review after rebase.

**Comment by jnewbery** [comment]
rebased

**Comment by dongcarl** [comment]
Took some time to prove to myself that there are no unintended behaviour changes. It wasn't obvious to me that 5e5d58c829001a774e2481a4a8c3598917b06e53 is trivially correct, so I'm most likely missing some context here.

My understanding is that prior to 5e5d58c829001a774e2481a4a8c3598917b06e53, `BlockChecked` is only called when `CheckBlock` or `AcceptBlock` returns false in `ProcessNewBlock`, causing `ProcessNewBlock` to return early right after calling `BlockChecked`.

However, it seems that after 5e5d58c829001a774e2481a4a8c3598917b06e53, it is possible for both `CheckBlock` and `AcceptBlock` to return true, but for `dos_state.IsValid()` to be false, causing `BlockChecked` to be called in `BlockProcessed` where it wouldn't have before. Perhaps this is fine and intended, but I just wanted to make sure.

Here's the diff I used for testing, and it seems to fail at `validation_block_tests/processnewblock_signals_ordering`:
```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 524ae94b9a..3f5737fd35 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3856,9 +3856,11 @@ bool ChainstateManager::ProcessNewBlock(const CChainParams& chainparams, const s
             ret = ::ChainstateActive().AcceptBlock(pblock, dos_state, chainparams, &pindex, fForceProcessing, nullptr, fNewBlock);
         }
         if (!ret) {
+            assert(!dos_state.IsValid());  // We should always trigger BlockChecked in BlockProcessed in this codepath
             return error("%s: AcceptBlock FAILED (%s)", __func__, dos_state.ToString());
         }
     }
+    assert(dos_state.IsValid());  // We should NOT trigger BlockChecked in BlockProcessed in this codepath
 
     NotifyHeaderTip();
 
```

**Comment by jnewbery** [comment]
The first commit has now been merged in #19533. I've rebased the remaining commits on master.

**Review by jonatack — COMMENTED** [review]
Almost ACK all commits except am still reviewing [removal of the call to BlockChecked() in PNB](https://github.com/bitcoin/bitcoin/pull/17479/files#diff-24efdb00bfbe56b140fb006b562cc70bL3861) in 372913c9 and other reasoning about this that @dongcarl states well in https://github.com/bitcoin/bitcoin/pull/17479#issuecomment-658293827. A few notes below. Code review, debug-built and ran tests at each commit; running a node on the branch ATM and observing behavior.

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
I'm focused on other things right now, so I'm going to close this with the intention of opening it again at some point in the future.
